packages/block-editor/src/components/preset-input-control/index.js

```diff
@@ -0,0 +1,367 @@
+import { settings } from '@wordpress/icons';
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	CUSTOM_VALUE_SETTINGS,
+	ICON_SIZE,
+	RANGE_CONTROL_MAX_SIZE,
+} from './constants';
+import {
+	getCustomValueFromPreset,
+	getPresetValueFromCustomValue,
+	getSliderValueFromPreset,
+	isValuePreset,
+} from './utils';
+
+/**
+ * PresetInputControl component for selecting preset values or entering custom values.
+ *
+ * Displays preset values as either a slider (for <= 8 presets) or a select dropdown.
+ * Allows toggling to custom value mode with a UnitControl and RangeControl.
+ * Handles unit tracking and conversion between preset and custom values.
+ *
+ * @param {Object}   props                     Component props.
+ * @param {boolean}  props.allowNegativeOnDrag Whether to allow negative values during drag operations.
+ * @param {string}   props.ariaLabel           Accessible label for the control.
+ * @param {string}   props.className           Optional CSS class name.
+ * @param {Object}   props.customValueSettings Optional custom value settings for max/steps per unit.
+ * @param {boolean}  props.disableCustomValues Whether to disable custom value input.
+ * @param {Object}   props.icon                Icon to display alongside the control.
+ * @param {boolean}  props.isMixed             Whether the current value is mixed (multiple values).
+ * @param {number}   props.minimumCustomValue  Minimum allowed custom value.
+ * @param {Function} props.onChange            Callback when value changes.
+ * @param {Function} props.onMouseOut          Callback for mouse out events.
+ * @param {Function} props.onMouseOver         Callback for mouse over events.
+ * @param {Function} props.onUnitChange        Callback when unit changes.
+ * @param {Array}    props.presets             Array of preset objects with name, slug, and size.
+ * @param {string}   props.presetType          Type of preset (e.g., 'spacing', 'border-radius').
+ * @param {string}   props.selectedUnit        Currently selected unit (e.g., 'px', 'em').
+ * @param {boolean}  props.showTooltip         Whether to show tooltip on custom UnitControl.
+ * @param {Array}    props.units               Array of available unit objects (can include max and step).
+ * @param {string}   props.value               Current value (preset or custom).
+ *
+ * @return {Element} The PresetInputControl component.
+ */
+export default function PresetInputControl( {
+	allowNegativeOnDrag = false,
+	ariaLabel,
+	className: classNameProp,
+	customValueSettings = CUSTOM_VALUE_SETTINGS,
+	disableCustomValues,
+	icon,
+	isMixed,
+	value,
+	minimumCustomValue,
+	onChange,
+	onMouseOut,
+	onMouseOver,
+	onUnitChange,
+	presets = [],
+	presetType,
+	selectedUnit,
+	showTooltip,
+	units,
+} ) {
+	// Treat value as a preset value if the passed in value matches the value of one of the presets.
+	value = getPresetValueFromCustomValue( value, presets, presetType );
```

Comment:
Is it okay to mutate the value here?  I mean, it works, just thinking of later debugging and stable props.

Thinking of something like

```js
const normalizedValue = useMemo( 
    () => getPresetValueFromCustomValue( value, spacingSizes ),
    [ value, spacingSizes ]
);
```

I could be way off!